nbgitpuller/plugins/plugin_helper.py

```diff
@@ -0,0 +1,146 @@
+import subprocess
+import os
+import stat
+import logging
+import requests
+from requests_file import FileAdapter
+import shutil
+import re
+
+
+# for large files from Google Drive
+def get_confirm_token(response):
+    for key, value in response.cookies.items():
+        if key.startswith('download_warning'):
+            return value
+    return None
+
+
+# sets up the a local repo that acts like a remote
+def initialize_local_repo(local_repo_path):
+    logging.info(f"Creating local_repo_path: {local_repo_path}")
+    try:
+        os.makedirs(local_repo_path)
+    except OSError:
+        logging.info(f"Directory exists: {local_repo_path}")
+
+    subprocess_helper("git init --bare", local_repo_path)
+
+    # Make our bare repository serveable over dumb HTTP
+    hook_path = os.path.join(local_repo_path, 'hooks', 'post-update')
+    os.rename(
+        os.path.join(local_repo_path, 'hooks', 'post-update.sample'),
+        hook_path
+    )
+    os.chmod(hook_path, os.stat(hook_path).st_mode | stat.S_IEXEC)
+
+
+# local repo cloned from the "remote" which is in user drive
+def clone_local_origin_repo(origin_repo_path, temp_download_repo):
+    logging.info(f"Creating temp_download_repo: {temp_download_repo}")
+    try:
+        os.makedirs(temp_download_repo)
+    except OSError:
+        logging.info(f"Directory exists: {temp_download_repo}")
+
+    cmd = f"git clone file://{origin_repo_path} {temp_download_repo}"
+    subprocess_helper(cmd, temp_download_repo)
+
+
+# this is needed to unarchive various formats(eg. zip, tgz, etc)
+def determine_file_extension(url, response):
+    file_type = response.headers.get('content-type')
+    content_disposition = response.headers.get('content-disposition')
```

Comment:
Can we just rely solely on content-type header instead?

Reply:
I am relying on the webserver to set the headers correctly. -- which is a nightmare. 
- Github(or random webserver with publicly accessible space) does not include content-type nor content-disposition so I extract the file extension from the URL
- Dropbox sets content-type to application/binary for gzip but includes the content-disposition so I parse the content-disposition for the filename
- Google includes content-type as well as content-disposition

I am actually not sure what to do here now. Google is the only one you can't determine the compression type from the file extension at the end of the URL but it seems dicey and brittle to rely on the URL to tell you the compression type anyway. 

Another option would be to read the first few bytes of the file and map that to the compression type -- but that also seems likely to produce a false positive at some point:
https://stackoverflow.com/questions/19120676/how-to-detect-type-of-compression-used-on-the-file-if-no-file-extension-is-spe

We could also just let the shutil.unpack_archive function give all the various compression types a shot until it works -- zip, gztar, etc. Something like this:
try:
    shutil.unpack_archive(temp_download_file, "/tmp", "zip")
except Exception:
    shutil.unpack_archive(temp_download_file, "/tmp", "gztar")

And finally, I guess we could force the user to include the compression type on the link puller URL but that seems like a hassle.

What do you think?

Reply:
I think we should make users think in terms of 'providers' instead. For example, if we use pluggy setuptools integration to discover plugins (https://github.com/jupyterhub/nbgitpuller/pull/194/files#r684300836), we can require that users pass in a 'provider' query param that matches the name of the plugin. For example, we know to call the google drive plugin if `provider` is set to `google-drive`, etc. This is important in the future for us to deal with Auth, and for working with other arbitrary providers (like GCS maybe?). I think compression should be handled by the provider as it knows what is necessary, so we don't have to try and detect this.

Reply:
ok that sounds good and answers another question I had -- do put the provider on the URL. Good deal. I will fix this up and bring it in

Reply:
So the plugin would receive all params passed in the query, and will make a determination on what it'll have to do. This means we can pass *anything* to 'url' in the future, and it's upto the provider to interpret it.

Reply:
See https://github.com/jupyterhub/the-littlest-jupyterhub/blob/master/docs/contributing/plugins.rst for how we do this in the Littlest JupyterHub. https://github.com/jupyterhub/the-littlest-jupyterhub/blob/master/tljh/hooks.py has the hooks, and you can see them be called from various places in the codebase